File: app/src/main/java/com/upsage/evosummermobilelab/data/intefaces/NoteDao.java
package com.upsage.evosummermobilelab.data.intefaces;

import androidx.paging.DataSource;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import com.upsage.evosummermobilelab.data.entries.Note;

@Dao
public interface NoteDao {

    @Query("SELECT * FROM note WHERE id = :noteId")
    Note loadById(int noteId);

    @Query("SELECT * FROM note WHERE description LIKE :first")
    DataSource.Factory<Integer, Note> findByDescription(String first);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    long[] insertAll(Note... users);

    @Delete
    void delete(Note id);

    @Update
    void update(Note note);

    @Query("select * from note order by update_date desc")
    DataSource.Factory<Integer, Note> getAllPagedDesc();

    @Query("select * from note order by update_date asc")
    DataSource.Factory<Integer, Note> getAllPagedAsc();

}

File: app/src/main/java/com/upsage/evosummermobilelab/data/intefaces/NoteDao.kt
package com.upsage.evosummermobilelab.data.intefaces

import androidx.paging.DataSource
import androidx.room.*
import com.upsage.evosummermobilelab.data.Note

@Dao
interface NoteDao {

    @get:Query("select * from note order by update_date desc")
    val allPagedDesc: DataSource.Factory<Int, Note>

    @get:Query("select * from note order by update_date asc")
    val allPagedAsc: DataSource.Factory<Int, Note>

    @Query("SELECT * FROM note WHERE description LIKE :first")
    fun findByDescription(first: String): DataSource.Factory<Int, Note>

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun insertAll(vararg users: Note): LongArray

    @Delete
    fun delete(id: Note)

    @Update
    suspend fun update(note: Note)

}